test/stmt/statements.swift

```diff
@@ -177,6 +177,7 @@ return 42 // expected-error {{return invalid outside of a func}}
 return // expected-error {{return invalid outside of a func}}
 
 func NonVoidReturn1() -> Int {
+  _ = 0
```

Comment:
I noticed that a bunch of tests have `_ = 0` added which is not ideal, if a single expression is left in the body is it going to produce different diagnostics? I think one way to work around this is to check whether a single expression is a declaration and avoid wrapping it in return in that case or wrap its result.

Reply:
Absolutely--this is not good.  

The [change](https://github.com/apple/swift/pull/23251/commits/e30511bdb8d8b4f0adc1b75e0db66992f4e25a74#diff-a5731a934e0c085bf4b84dabd27a6ffdR5719) I just pushed corrects a number of these.  Is this what you had in mind in order to "check whether a single expression is a declaration"?  

```
if (auto SE = dyn_cast<SequenceExpr>(E->getSemanticsProvidingExpr())) {
  if (SE->getNumElements() > 0 && isa<AssignExpr>(SE->getElement(1))) {
    // This is an assignment.  We don't want to implicitly return 
    // it.
    return;
  }
}
```

It does look like a few of them are appropriate:
(1) `test/Generics/requirement_inference.swift` -- we don't want to implicitly return the call to x.p1() which returns void -- dataflow analysis would detect an error (that nothing is returned) but that's not run as part of the test
(2) `test/Sema/availability_versions.swift` -- we don't want to implicitly return the call to `functionAvailableOn10_51` which would happen if the body consisted only of that
(3) `validation-test/stdlib/AssertDiagnosticsSIL.swift` -- we don't want to implicitly return the call to `assertionFailure` which would happen if the body consisted only of that
(4) `test/SILGen/super.swift` -- we don't want the body to consist of only a call to super.classMethod() since the function is expected to return Self and super.classMethod() returns Void


The change in `test/stmt/statements.swift` is questionable and corresponds to a questionable part of the patch: https://github.com/apple/swift/pull/23251/files#diff-a5731a934e0c085bf4b84dabd27a6ffdR5708 .  In the code I linked, a () result is being added to ReturnStmt.  Something like this is being done for closures, so I had it commented out to begin with but over the course of the review it was commented in.  In view of this test failure, it seems like that code should go.  Is that right?

I'm looking into `test/decl/subscript/subscripting.swift` and `test/decl/var/static_var.swift` further to see whether it makes sense to change those tests.

Reply:
@nate-chandler Yes, something like that. This got me thinking, maybe we should even attempt that. Because as in cases you have mentioned, I think it makes sense make it not `_ = 0` but something reasonable, under new rules implicit return is a correct behavior. All of the single statement functions would have an implicit return added to them if the result type of the function is specified. It maybe makes sense to add tailored diagnostic for that case though which says that single expression function can't implicitly return T because result type is specified as U.
And in case of assignments say that `non-void function should return a value`.

Reply:
@xedin To make sure I understand, the idea is that:

(1) Assignments should be implicitly returned.  This
```swift
func foo() -> Int {
    a = b
}
```
should be rewritten as
```swift
func foo() -> Int {
    return a = b
}
```

(2) There should be a diagnostic for single expression return of ().  This test should pass:
```swift
func foo() -> Int {
    a = b // expected-error {{ non-void function should return a value }}
}
```

(3) There should be a special diagnostic for the return of an implicit ().  This code should remain:  
```swift
if (!returnStmt->hasResult()) {
    auto returnExpr = TupleExpr::createEmpty(Context,
                                             SourceLoc(),
                                             SourceLoc(),
                                             /*implicit*/true);
    returnStmt->setResult(returnExpr);
    AFD->setHasSingleExpressionBody();
    AFD->setSingleExpressionBody(returnExpr);
  }
}
```
This code
```swift
func foo() -> Int {
    return
}
```
should still be rewritten as
```swift
func foo() -> Int {
    return ()
}
```
And this test should pass:
```swift
func foo() -> Int {
    return // expected-error {{ non-void function should return a value }}
}
```

(4) There should be a diagnostic for incorrect implicit return.  This test should pass:
```swift
func foo() -> String {
    17 // expected-error {{cannot convert implicitly returned single expression of type 'Int' to return type 'String'}}
}
```

We should continue to insert () as the result of return in 

(5) The tests that are modified to have a second element so as not to have an implicit return inserted (as at least `test/Generics/requirement_inference.swift`, `test/Sema/availability_versions.swift`, `validation-test/stdlib/AssertDiagnosticsSIL.swift`, and `test/SILGen/super.swift` will need to be) should use an expression other that `_ = 0`?  For example, 

I'm drawing (5) from what you wrote here

> Because as in cases you have mentioned, I think it makes sense make it not _ = 0 but something reasonable, under new rules implicit return is a correct behavior. 

but I'm not sure I'm understanding correctly.

Thank you!